dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,407 @@
+
+Xous is a microkernel-based OS written in Rust, currently in Alpha state. Xous
+is used as a secure microkernel for [Betrusted](betrusted.io) device and is
+focused on RISC-V, which right now is the only architecture supported. However,
+x86 already contains
+[dummy implementation](https://github.com/betrusted-io/xous-core/blob/30b82b25b100e958790973c129dc49e1acca79ec/kernel/src/arch/x86_64.rs)
+which probably will be extended someday.
+
+
+#### Running in DLME
+
+Xous currently cannot run on x86, and DLME on RISC-V is not covered here. We
+have opened an [issue](https://github.com/betrusted-io/xous-core/issues/139)
+raising the subject, among others, about x86 support.
+
+#### Fobnail integration
+
+Xous has network support but currently lacks USB host drivers and TPM support.
+Also, it lacks C interface for calling the kernel and lacks C library. Getting
+[fobnail-attester](https://github.com/fobnail/fobnail-attester) running may
+require a significant amount of work. Also, Xous needs to gain kexec-like
+abilities to chainload target OS.
+
+### seL4
+
+seL4 is a secure L4 family microkernel written in C. It has strong security
+guarantees assured by [Formal proofs](https://sel4.systems/Info/FAQ/proof.pml),
+however these proofs are still incomplete for x86, see
+[Supported Platforms](https://docs.sel4.systems/Hardware) for an up-to-date
+verification status. seL4 due to its microkernel nature provides higher
+isolation, breach in one of the components (like USB driver, network stack)
+wouldn't compromise entire OS contrary to monolithic kernels.
+
+We have opened an [issue](https://github.com/seL4/seL4/issues/832) about seL4
+usage as secure bootloader.
+
+#### Running in DLME
+
+It should be runnable, however due to microkernel nature all programs run in
+unprivileged mode which complicates booting of the target OS. seL4 would have to
+gain kexec-like abilities.
+
+#### Fobnail integration
+
+seL4 provides virtually no drivers, except a few drivers listed here.
+According to the page linked above, `libusbdrivers` is inactive and lacks XHCI
+support. seL4 has basic support for an old version of musl C library (v1.1.16).
+
+Using seL4 would require a significant amount of work:
+- Extending USB drivers
+- Implementing USB EEM driver
+- Implementing TPM driver
+- Porting [fobnail-attester](https://github.com/fobnail/fobnail-attester)
+- seL4 runs all processes in unprivileged mode, so the kernel itself would have
+  to be modified to allow booting of another kernel
+
+[Genode](https://github.com/genodelabs/genode) is an framework for creating
+custom, specialized OSes. It provides USB drivers (including `usb-net` driver
+from which should handle USB EEM) and a network stack. Its support for seL4 used
+to be incomplete, and many components were broken. However, that might have
+changed, and Genode may be an easier way to get software running on seL4. We
+have opened an [issue](https://github.com/genodelabs/genode/issues/4480) here
+too.
```

Comment:
@pietrushnic asked me to review this PR. Disclaimer: I'm not affiliated with Genodelabs. However, we at Gapfruit have been working with the Genode Framework for almost a decade. In my view, Genode is highly under-appreciated. So let me try to sum up why I believe Genode deserves a more prominent role in this document than merely a sub-section of seL4. Genode is a framework to design component-based scenarios - or whole operating systems - on top of various microkernels. Every component is strongly isolated and only gets access to resources and services in an explicit manner. The microkernel guarantees strong isolation and maintains access control to various resources via capability objects. 

The system designer defines the component topology in a declarative and nested manner - making it possible to abstract sub-systems while maintaining the possibility to define fine-grained trust relationships between the components. 

One example scenario that illustrates this topology quite well is the [lighttpd](https://github.com/genodelabs/genode/blob/master/repos/ports/run/lighttpd.run) run script. This script creates a static boot image with only the necessary components to serve a webpage with lighttpd: `timer`, `nic_drv`, `nic_router`, and `lighttpd`. This run script works (mostly) out of the box with different kernels (seL4, nova, base-hw, linux etc) on different hardware (ARM, x86, RISC-V). This example illustrates quite well what a component topology for the fobnail project could look like. It also demonstrates the possibility of porting POSIX components to their FreeBSD-based libc port.

While the lighttpd run script shows how a static scenario can look like, more dynamic systems like the [Sulpt OS](https://genode.org/documentation/articles/sculpt-21-10) or the headless Gapfruit OS are possible. 

Regarding drivers, I believe that most required drivers already work (USB host driver, USB NIC driver, ACPI etc). We at Gapfruit also have ported WolfTPM, which we haven't open-sourced yet.

One missing requirement is the possibility to `kexec` from a Genode scenario to another operating system. That would be a great research topic. I can see two intermediate solutions until then:
- On x86_64 Genode has ported VirtualBox on top of the nova micro-hypervisor.
- When running a Genode scenario on Linux, each component is isolated with Seccomp and the capability objects are represented with file-descriptors. This would make it possible to create a minimal Linux kernel that would kexec the main OS after a successful integrity check.

Reply:
What advantage does Genode have over CAmkES?  Genode seems like a poor choice for a static system like this.

Reply:
> What advantage does Genode have over CAmkES? 

It's been a while since I played with CAmkES, so my judgment might be outdated. However, from an engineering perspective, I would say Genode is way ahead in terms of tooling, supported drivers, supported driver environments, supported runtimes, hardware, and maturity.

> Genode seems like a poor choice for a static system like this.

How so? In the beginning, Genode was static as well. On Genode, you don't pay for what you don't use. That also counts for _not_ using the dynamic features of Genode. Do you mean from a security perspective? On paper, SeL4 + CAmkES indeed sounds compelling. Formal verification _does_ make a strong argument. IMHO Genode takes a more reasonable approach. The declarative service topology makes it possible to identify the components that are critical to the system. These vital components can then be made as resilient and safe as possible - using various techniques such as formal verification, static analysis, fuzzing, profiling, safe languages, etc. Obviously, the kernel _is_ critical. But so are e.g. multiplexers or components that have access to secrets (similar to split GPG in Qubes). 

From a marketing perspective, the formal verification argument might make sense. In this regard, it is worth considering the Nova hypervisor for x86_64. AFAIK Nova is also partly formally verified. 

What I forgot to mention: On x86_64, using Genode combined with Nova, IOMMU-protection is fully implemented and running in production.

That said, I may don't see all requirements and implications of using Genode for this project. CAmkES may very well be the best option. And, from my perspective, definitely better than some of the other alternatives mentioned in this document. 

@DemiMarie I'm eager to learn why you think CAmkES is a better fit than Genode.

Reply:
> > What advantage does Genode have over CAmkES?
> 
> It's been a while since I played with CAmkES, so my judgment might be outdated. However, from an engineering perspective, I would say Genode is way ahead in terms of tooling, supported drivers, supported driver environments, supported runtimes, hardware, and maturity.

That does not surprise me, though the seL4 Core Platform might be better once it is implemented.

> > Genode seems like a poor choice for a static system like this.
> 
> How so? In the beginning, Genode was static as well. On Genode, you don't pay for what you don't use. That also counts for _not_ using the dynamic features of Genode.

I was not aware of this.

> Do you mean from a security perspective?

That is where basically all of my arguments come from, so yes.  I have never used any of these frameworks myself.

> On paper, SeL4 + CAmkES indeed sounds compelling. Formal verification _does_ make a strong argument. IMHO Genode takes a more reasonable approach. The declarative service topology makes it possible to identify the components that are critical to the system. These vital components can then be made as resilient and safe as possible - using various techniques such as formal verification, static analysis, fuzzing, profiling, safe languages, etc. Obviously, the kernel _is_ critical. But so are e.g. multiplexers or components that have access to secrets (similar to split GPG in Qubes).

That is indeed the case.  In fact, this OS needs support for `kexec()`, so at least the root task will be as critical as the kernel.

> From a marketing perspective, the formal verification argument might make sense.

Why “from a marketing perspective”?  To me, formal verification is a substantial security win, though I would be interested to hear from others, such as yourself and @lsf37.

> In this regard, it is worth considering the Nova hypervisor for x86_64. AFAIK Nova is also partly formally verified.

Can you link to the proofs?

> What I forgot to mention: On x86_64, using Genode combined with Nova, IOMMU-protection is fully implemented and running in production.

That is a definite plus.  Is interrupt remapping included?

> That said, I may don't see all requirements and implications of using Genode for this project. CAmkES may very well be the best option. And, from my perspective, definitely better than some of the other alternatives mentioned in this document.

> @DemiMarie I'm eager to learn why you think CAmkES is a better fit than Genode.

A few reasons:

1. Genode’s licensing terms (AGPLv3 or commercial) may be a dealbreaker.  In particular, they are incompatible with the licenses of Heads and of Qubes OS.  Additionally, @pietrushnic might find the AGPLv3 too restrictive.  As this is an open source project, the commercial license is not applicable.
2. Genode does not have an assurance story, and AFIAK no intention of developing one.
3. Genode’s approach of abstracting the kernel in use seems like it would prevent using the full feature set of whatever kernel is chosen for the project.
4. My understanding is that Genode requires components to be written in C++, which is not a language I would choose here.  I could obviously be wrong about this, though.

Reply:
In embedded realm anything containing *GPLv3 is no go.

Reply:
> That is indeed the case. In fact, this OS needs support for kexec(), so at least the root task will be as critical as the kernel.

To avoid confusion, the `kexec` syscall would need to be implemented in kernel space e.g. SeL4. And there would be a user-space component that calls `kexec`. I assume that is the component that you referred to as root task. On Genode this component would probably be a hybrid component. More about that further down.

> Why “from a marketing perspective”? To me, formal verification is a substantial security win, though I would be interested to hear from others, such as yourself and @lsf37.

That is a bad choice of words from my side. I was wrong to start talking about marketing in this technical discussion. In no way do I want to discredit formal verification. I tried to formulate its value in the "security perspective" section above. Formal verification is just very expensive from an engineering perspective.

> > What I forgot to mention: On x86_64, using Genode combined with Nova, IOMMU-protection is fully implemented and running in production.

> That is a definite plus. Is interrupt remapping included?

Yes. AFAIK interrupts are remapped regardless of IOMMU protection: 
- https://genode.org/documentation/genode-foundations/21.05/architecture/Core_-_the_root_of_the_component_tree.html
- https://genode.org/documentation/genode-foundations/21.05/under_the_hood/Execution_on_the_NOVA_microhypervisor_(base-nova).html#IOMMU_support

> 1. Genode’s licensing terms (AGPLv3 or commercial) may be a dealbreaker. In particular, they are incompatible with the licenses of Heads and of Qubes OS. Additionally, @pietrushnic might find the AGPLv3 too restrictive. As this is an open source project, the commercial license is not applicable.

I understand how this might be a challenge.

> 2. Genode does not have an assurance story, and AFIAK no intention of developing one.

I don't understand what this means. Do you mean for a certification body?

> 3. Genode’s approach of abstracting the kernel in use seems like it would prevent using the full feature set of whatever kernel is chosen for the project.

You are correct. When developing Genode components, the abstractions you'd use are services, not syscalls. However, Genode has the notion of hybrid components, e.g., on Linux, to interact with the Linux filesystem: 
- https://github.com/genodelabs/genode/tree/master/repos/os/src/server/lx_fs

It seems like the only kernel feature missing for this project is `kexec`. For a rapid prototype using Linux as kernel, a hybrid component could issue this syscall. At the same time, other functionality such as integrity checks, TPM drivers, etc. are implemented as regular Genode components. Once, e.g., some form of `kexec` is available on SeL4 or NOVA, you'd only have to rewrite this specific hybrid component.

> 4. My understanding is that Genode requires components to be written in C++, which is not a language I would choose here. I could obviously be wrong about this, though.

Partly true. Most components are written in C++. However, they do not rely on any libc or stdlib. These libs are linked to the component only when needed. Critical components are designed to keep the TCB/LOC as low as possible and thus are written without these libs. More about the rationale behind this topic here: https://genodians.org/nfeske/2019-01-22-conscious-c++
Btw, ADA/SPARK is also supported. And there is an (abandoned) port for Rust.

Reply:
@DemiMarie goal of this document is to make people aware of various options. We would also try to have rough evaluation of effort, for potential future sponsors information. If there will be enough interest in Genode as DLME OS, we definitely will be open for using it. I don't see Fobnail Project ecosystem as relying only on one DLME OS. 
AGPLv3 may be deal breaker for embedded environment, but maybe for other use cases this license is not an issue.

Reply:
@pietrushnic I updated Genode section.

Reply:
@sidhussmann any chance you can look into that? From my perspective we gathered quite a lot of insights, but this conversation was started by you and I would like to make sure you are ok with closing it.

Reply:
@pietrushnic I'm fine with the current state.

Reply:
@sidhussmann thanks for confirmation. So I will resolve this thread.

Reply:
@sidhussmann thank you for your contribution again. Feel free to reach us out when you will need our support.